### PR TITLE
feat(Component-relationship): add modal to define component relationship

### PR DESCRIPTION
**Pull request by sahil143**
## Fixes 

https://issues.redhat.com/browse/HAC-5432

## Description

add component relation modal for applications

TODO:
- [ ] add unit tests 

## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Screen shots / Gifs for design review 

Kapture 2024-01-30 at 11 14 22

## Browser conformance: 

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by MariaLeonova** [comment]
@karthikjeeyar yes, I think if there's only one, there's no need for the Define relationships option to be active

File: src/components/ComponentRelation/useNudgeData.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            return [...acc, data];
+          }
+          if (val.status?.[NudgeStats.NUDGED_BY]) {
+            const data: ComponentRelationValue = {
+              source: val.metadata.name,
+              nudgeType: ComponentRelationNudgeType.NUDGED_BY,
+              target: val.status?.[NudgeStats.NUDGED_BY],
+            };
+            return [...acc, data];
+          }
```
[the remark this answers is not in the document]
Comment (arewm) [reply src/components/ComponentRelation/useNudgeData.ts:28]: I agree with @MariaLeonova . If possible, we should only show the relationships that users set and not the auto-generated inverse ones.

[the remark this answers is not in the document]
Comment (sahil143) [reply src/components/ComponentRelation/useNudgeData.ts:28]: @MariaLeonova  If user creates `is nudged by` relationship then also we make a nudges API call behind the scenes. So if user creates a `is nudged by` relationship and then comes back to this modal after saving they will only see `nudges` relationships in the modal.

File: src/components/ComponentRelation/ComponentRelationshipDropdowns.tsx [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+};
+
+const SelectComponentsDropdown: React.FC<SelectComponentsDropdownProps> = ({
+  children,
+  toggleText,
+  onSelect,
+  closeOnSelect,
+  badgeValue,
+}) => {
+  const [isOpen, setIsOpen] = React.useState(false);
```
Comment (abhinandan13jan) [inline src/components/ComponentRelation/ComponentRelationshipDropdowns.tsx:29]: type boolean

File: src/components/ComponentRelation/cr-modals.tsx [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+const SuccessIcon: React.FC = () => (
+  <Icon status="success">
+    <CheckCircleIcon />
+  </Icon>
+);
+
+type ConfirmSubmissionComponentRelationModalProps = Pick<RawComponentProps, 'modalProps'>;
+
+export const ConfirmSubmissionComponentRelationModal: React.FC<
```
Comment (abhinandan13jan) [inline src/components/ComponentRelation/cr-modals.tsx:127]: can we move this out to separate file

**Comment by sahil143** [comment]
> There is no going back if I accidentally clicked on the add another relationship button, there is no remove button so I had to close the current process and start it again.

@MariaLeonova can you please take a look at this comment?

File: src/components/ComponentRelation/useNudgeData.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  const { namespace } = useWorkspaceInfo();
+  const [components, loaded, error] = useComponents(namespace, application);
+  const nudgeData: ComponentRelationValue[] = React.useMemo(() => {
+    return loaded && !error
+      ? components.reduce((acc, val: ComponentKind) => {
+          if (val.spec?.[NudgeStats.NUDGES]) {
+            const data: ComponentRelationValue = {
+              source: val.metadata.name,
+              nudgeType: ComponentRelationNudgeType.NUDGES,
+              target: val.spec[NudgeStats.NUDGES],
```
Comment (karthikjeeyar) [inline src/components/ComponentRelation/useNudgeData.ts:17]: we can take into account that only available components in the application will be used to show relationships.

```suggestion
              target: val.spec[NudgeStats.NUDGES]?.filter(
                (cname) => !!components?.find((co) => co?.metadata?.name === cname),,
```

Comment (arewm) [reply src/components/ComponentRelation/useNudgeData.ts:17]: This restriction to limit nudges within an Application have actually been loosened. We should no longer filter to only Components in the one Application. 

@MariaLeonova , this has a potential adverse effect on the UX as the number of components can be much larger. 

Would it be reasonable/possible to be able to filter components based on their Application? Or to have Components listed with their Application name (i.e. `<Application>/<Component>`) for grouping "similar" Components together?

Comment (sahil143) [reply src/components/ComponentRelation/useNudgeData.ts]: @arewm We will add support for component relations across applications in a new story.

**Review by rohitkrai03 — APPROVED** [review]
/approve

File: src/components/ComponentRelation/useNudgeData.ts (hunk 4 above)
Comment (arewm) [reply src/components/ComponentRelation/useNudgeData.ts]: @sahil143 , do you have the feature to for cross-Application support?

**Pull request merged**